src/pkg/packager/update.go

```diff
@@ -0,0 +1,149 @@
+		return fmt.Errorf("failed to create update: %w", err)
+	}
+
+	if err := os.WriteFile(packageConfigFile, []byte(updatedZarfYaml), helpers.ReadAllWriteUser); err != nil {
+		return fmt.Errorf("failed to write updated %s: %w", packageConfigFile, err)
+	}
+
+	l.Info("successfully updated images", "path", packageConfigFile)
+	return nil
+}
+
+func createUpdate(zarfPackage v1alpha1.ZarfPackage, imagesScans []ComponentImageScan, astFile *ast.File) (string, error) {
+	// Note: yamlpath support of goccy/go-yaml only has index-based lookup
+	componentToIndex := make(map[string]int, len(zarfPackage.Components))
+	for i, component := range zarfPackage.Components {
+		componentToIndex[component.Name] = i
+	}
+
+	for _, scan := range imagesScans {
+		if len(scan.Matches)+len(scan.PotentialMatches)+len(scan.CosignArtifacts) == 0 {
+			continue
+		}
+
+		componentIndex, exists := componentToIndex[scan.ComponentName]
+		if !exists {
+			continue
+		}
+
+		allImages := slices.Concat(scan.Matches, scan.PotentialMatches, scan.CosignArtifacts)
+
+		componentMerge := map[string]any{
+			"images": allImages,
+		}
+		componentNode, err := yaml.ValueToNode(componentMerge, yaml.IndentSequence(true))
+		if err != nil {
+			return "", fmt.Errorf("failed to create YAML node for component %s: %w", scan.ComponentName, err)
+		}
+
+		path, err := yaml.PathString(fmt.Sprintf("$.components[%d]", componentIndex))
+		if err != nil {
+			return "", fmt.Errorf("failed to create YAML path for component %s: %w", scan.ComponentName, err)
+		}
+
+		if err := path.MergeFromNode(astFile, componentNode); err != nil {
+			return "", fmt.Errorf("failed to merge images for component %s: %w", scan.ComponentName, err)
+		}
+	}
+
+	return astFile.String(), nil
+}
+
+func updateNeeded(zarfPackage v1alpha1.ZarfPackage, imageScans []ComponentImageScan) bool {
+	scanMap := make(map[string]map[string]struct{}, len(imageScans))
+
+	for _, scan := range imageScans {
+		combined := slices.Concat(scan.Matches, scan.PotentialMatches, scan.CosignArtifacts)
+		imageSet := make(map[string]struct{}, len(combined))
+		for _, img := range combined {
+			imageSet[img] = struct{}{}
+		}
+		scanMap[scan.ComponentName] = imageSet
+	}
+
+	for _, component := range zarfPackage.Components {
+		imageSet, found := scanMap[component.Name]
+		if !found {
+			return true
+		}
+
+		for _, img := range component.Images {
+			if _, found := imageSet[img]; !found {
+				return true
+			}
+		}
+	}
+
+	componentMap := make(map[string]map[string]struct{}, len(zarfPackage.Components))
+	for _, component := range zarfPackage.Components {
+		imageSet := make(map[string]struct{}, len(component.Images))
+		for _, img := range component.Images {
+			imageSet[img] = struct{}{}
+		}
+		componentMap[component.Name] = imageSet
+	}
+
+	for _, scan := range imageScans {
+		componentImages, found := componentMap[scan.ComponentName]
+		if !found {
+			return true
+		}
+
+		combined := slices.Concat(scan.Matches, scan.CosignArtifacts)
```

Comment:
Not sure if leaving out `potentialMatches` was intentional, but I think we should keep them in. Feels more consistent since we always add them when we do an update